### PR TITLE
feat: Repository APIs to allow developer a consistent interface to a trestle repo.

### LINKED ISSUE #544: As a software developer I would like a consistent interface to the trestle repository.
## Issue description / feature objectives
Issue built in discussion with @gregelin. The current abstractions for interacting with the repository are based primarily around the initial 'actions' that a user may want to take on the CLI.

As a developer using a trestle project I want to interact with the concept of a 'project' or repository. E.g. we want to be able to update and retrieve objects in a simple way such as this:

```python
import pathlib
from trestle.oscal import catalog
from trestle.oscal import profile


class Repository():

    def __init__(self, root_dir: pathlib.Path) -> None:
        pass

    def get_catalog(name: str) -> catalog.Catalog:
        pass

    def write_catalog(name: str, profile: profile.Profile, overwrite=False, distribute=True):
        pass

```
where root_dir would be the centralised project directory as a `pathlib.Path`. This would allow the centralisation of some of the pain that is currently endured in the CLI classes allowing a progressive refactor.

Open questions:
1) The way the current implementation is portrayed above it leaps from the repository to the concrete `OscalBaseModel`. To keep the interface clean this might limit the actions performed by the repository class. Another option is to have the repo class veneer another abstraction.. eg,.

```python
class ManagedOSCAL:
    def __init__(self, root_dir: pathlib.Path, model_type: Type[OscalBaseModel], name: str) -> None:
        pass
    def read(self) -> OscalBaseModel:
        pass
    def write(self, model: OscalBaseModel, preserve_split: bool) -> None:
        pass
    def split(self, elements: List[str]) -> None:
        pass
    def merge(self, elements: List[str]) -> None:
        pass
    def add(self, element: str) -> None:
        pass
    def remove(self, element: str) -> None:
        pass
    def validate(self, mode: str) -> None:
        pass

class Repository:
    def __init__(self, root_dir: pathlib.Path) -> None:
        pass
    def import_model(self, model: OscalBaseModel, name: str) -> ManagedOSCAL:
        pass
    def list_models(self, model_type: Type[OscalBaseModel]) -> List[str]:
        pass
    def get_model(self, model_type: Type[OscalBaseModel], name: str) -> ManagedOSCAL:
        pass
    def delete_model(self, model_type: Type[OscalBaseModel], name: str) -> None:
        pass
    def assemble_model(self, model_type: Type[OscalBaseModel], name: str) -> None:
        pass
    def validate_model(self, model_type: Type[OscalBaseModel], name: str, mode: str) -> None:
        pass
```

```python
repo = Repository(my_directory)
component_defs: str = repo.list_models(trestle.oscal.component.ComponentDefinition)

assert component_defs[0] == 'rhel'

rhel_cd = repo.get_model(trestle.oscal.component.ComponentDefinition, 'rhel).read()

rhel_components = rhel_cd.components
```

## Completion Criteria
- [ ] Complete final design and document
- [ ] Implement such that content can be pulled and written back into the repository (assume no splitting is required)
- [ ] Write up issues for secondary functionaliy.
- [ ] Examine whether CLI classes should be refactored.

### PR DESCRIPTION
**Pull request by vikas-agarwal76**
## Types of changes

- \[ \] Bug fix (non-breaking change which fixes an issue)
- \[X \] New feature (non-breaking change which adds functionality)
- \[X \] Breaking change (fix or feature that would cause existing functionality to change)
- \[X \] My code follows the code style of this project.
- \[X \] My change requires a change to the documentation.
- \[ \] I have updated the documentation accordingly.
- \[ \] I have added tests to cover my changes.
- \[ \] All new and existing tests passed.
- \[ \] All commits are signed-off.

closes #544

Added API for main Repository class. API implementation for ManagedOSCAL is still pending. Added trestle_root args in the base command and its default value is set to Path.cwd().

### CODE REVIEW

**Comment by butler54** [comment]
@vikas-agarwal76 - can you update to the latest version in develop and then I can have a look?

**Comment by vikas-agarwal76** [comment]
@butler54 Updated latest from develop

**Comment by butler54** [comment]
Looking good so far. Please remove the commented out code.

**Comment by vikas-agarwal76** [comment]
@butler54 Pushed final changes to feat/repo. Please check. Also I see some tests failing. Pls. check that also.

**Comment by vikas-agarwal76** [comment]
@butler54 I have merged the latest from develop and the changes to test cases to handle trestle_root. Pls. review and merge into develop.

**Review by butler54 — CHANGES_REQUESTED** [review]
See the changes requested:
One is structural (location of files)
Many are commented out code.

File: trestle/utils/fs.py [hunk 1]
```diff
@@ -102,6 +102,7 @@ def has_trestle_project_in_path(path: pathlib.Path) -> bool:
 def get_contextual_model_type(path: pathlib.Path = None) -> Tuple[Type[OscalBaseModel], str]:
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/utils/fs.py:102]: Even if the user is providing a path then the check 
if not path.exists():
            path = pathlib.Path.cwd() / path

still needs to be there of check is path is valid. 

So we can leave the code as it is.

File: trestle/core/commands/assemble.py [hunk 2]
```diff
@@ -142,20 +141,27 @@ def assemble_model(cls, model_alias: str, object_type: Type[TLO], args: argparse
...
-            logger.error(f'Current working directory {Path.cwd()} is not the top level trestle project directory.')
+
+        # comment trestle_root = fs.get_trestle_project_root(Path.cwd())
+        # comment if not trestle_root:
+        # comment    logger.error(f'Current working directory {Path.cwd()} is not with a trestle project.')
+        # comment    return 1
+        # comment if not trestle_root == Path.cwd():
+        # comment    logger.error(f'Current working directory {Path.cwd()} is not the top level trestle project'
+        # comment                 + ' directory.')
+        # comment    return 1
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/assemble.py:152]: Done

File: trestle/core/commands/assemble.py [hunk 3]
```diff
@@ -142,20 +141,27 @@ def assemble_model(cls, model_alias: str, object_type: Type[TLO], args: argparse
...
+        trestle_root = args.trestle_root  # trestle root is set via command line in args. Default is cwd.
+        if not trestle_root or not fs.is_valid_project_root(args.trestle_root):
+            # comment logger.error(f'Current working directory {Path.cwd()} is not with a trestle project.')
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/assemble.py:155]: done

File: trestle/core/commands/import_.py [hunk 4]
```diff
@@ -66,10 +66,12 @@ def _run(self, args: argparse.Namespace) -> int:
             return 1
 
         # 1.2 Bad working directory if not running from current working directory
-        cwd = pathlib.Path.cwd()
-        trestle_root = fs.get_trestle_project_root(cwd)
-        if trestle_root is None:
-            logger.error(f'Current working directory: {cwd} is not within a trestle project.')
+        # comment cwd = pathlib.Path.cwd()
+        # comment trestle_root = fs.get_trestle_project_root(cwd)
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/import_.py:70]: done

File: trestle/core/commands/init.py [hunk 5]
```diff
@@ -38,14 +37,18 @@ class InitCmd(CommandPlusDocs):
     def _run(self, args: argparse.Namespace) -> int:
         """Create a trestle project in the current directory."""
         log.set_log_level_from_args(args)
-        dir_path = os.getcwd()
+        # comment dir_path = os.getcwd()
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/init.py:40]: done

File: trestle/core/commands/replicate.py [hunk 6]
```diff
@@ -161,10 +160,12 @@ def replicate_object(cls, model_alias: str, object_type: Type[TLO], args: argpar
         logger.debug('Entering replicate_object.')
 
         # 1 Bad working directory if not running from current working directory
-        cwd = pathlib.Path.cwd()
-        trestle_root = fs.get_trestle_project_root(cwd)
-        if trestle_root is None:
-            logger.error(f'Current working directory: {cwd} is not within a trestle project.')
+        # comment cwd = pathlib.Path.cwd()
+        # comment trestle_root = fs.get_trestle_project_root(cwd)
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/replicate.py:164]: done

File: trestle/core/commands/repo/trestle_repo.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
+# -*- mode:python; coding:utf-8 -*-
+
+# Copyright (c) 2020 IBM Corp. All rights reserved.
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/repo/trestle_repo.py:3]: done

File: trestle/core/commands/task.py [hunk 8]
```diff
@@ -64,9 +64,11 @@ def _run(self, args: argparse.Namespace) -> int:
             logger.error('Either a trestle task or "-l/--list" shoudl be passed as input arguments.')
             return 1
         # Ensure trestle directory (must be true)
-        trestle_root = fs.get_trestle_project_root(pathlib.Path.cwd())
-        if not trestle_root:
-            logger.error(f'Current working directory {pathlib.Path.cwd()} is not with a trestle project.')
+        # comment trestle_root = fs.get_trestle_project_root(pathlib.Path.cwd())
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/task.py:67]: done

File: trestle/utils/fs.py [hunk 9]
```diff
@@ -102,6 +102,7 @@ def has_trestle_project_in_path(path: pathlib.Path) -> bool:
...
     """Get the full contextual model class and full jsonpath for the alias based on the contextual path."""
     logger.debug(f'get contextual model type for input path {path}')
+    # FIXME: set default value of path to Path.cwd()
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/utils/fs.py:105]: done

File: trestle/utils/fs.py [hunk 10]
```diff
@@ -354,15 +356,18 @@ def get_contextual_file_type(path: pathlib.Path) -> FileContentType:
...
 
 
-def get_models_of_type(model_type: str) -> List[str]:
+def get_models_of_type(model_type: str, root: pathlib.Path) -> List[str]:
     """Get list of model names for requested type in trestle directory."""
     if model_type not in const.MODEL_TYPE_LIST:
         raise err.TrestleError(f'Model type {model_type} is not supported')
     # search relative to project root
-    trestle_root = get_trestle_project_root(pathlib.Path.cwd())
+    # comment trestle_root = get_trestle_project_root(pathlib.Path.cwd())
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/utils/fs.py:364]: done

File: trestle/core/commands/repo/trestle_repo.py [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+        repo = Repository(self.root_dir)
+        success = repo.validate_model(self.model_type, self.model_name)
+        return success
+
+
+class Repository:
+    """Repository object for performing operations on Trestle repository."""
+
+    def __init__(self, root_dir: pathlib.Path) -> None:
+        """Initialize trestle epository object."""
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/repo/trestle_repo.py:201]: done

File: trestle/core/commands/repo/trestle_repo.py [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+            raise TrestleError(f'Provided root directory {str(root_dir)} is not a valid Trestle root directory.')
+        self.root_dir = root_dir
+
+    def import_model(self, model: OscalBaseModel, name: str, content_type='json') -> ManagedOSCAL:
+        """Import OSCAL object into trestle repository."""
+        logger.debug(f'Importing model {name} of type {model.__class__.__name__}.')
+        model_alias = classname_to_alias(model.__class__.__name__, 'json')
+
+        # Work out output directory and file
+        plural_path = fs.model_type_to_model_dir(model_alias)
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/repo/trestle_repo.py:212]: Done.

File: trestle/core/commands/repo/__init__.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# -*- mode:python; coding:utf-8 -*-
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/repo/__init__.py:1]: done. moved trestle_repo.py to trestle/core/repository.py

File: trestle/core/commands/repo/trestle_repo.py [hunk 14]
```diff
@@ -0,0 +1,356 @@
+# -*- mode:python; coding:utf-8 -*-
```
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/repo/trestle_repo.py:1]: done. moved trestle_repo.py to trestle/core/repository.py

**Comment by vikas-agarwal76** [comment]
@butler54 Made all the changes suggested by you. See my reply to you individual comments. Also moved the trestle_repo.py to trestle/core/repository.py

File: trestle/core/commands/repo/trestle_repo.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (vikas-agarwal76) [reply trestle/core/commands/repo/trestle_repo.py]: Applied type checking to all relevant functions.

File: trestle/utils/fs.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (butler54) [reply trestle/utils/fs.py:102]: Okay awesome.

File: trestle/utils/fs.py [hunk 15]
```diff
@@ -154,6 +155,7 @@ def get_stripped_contextual_model(path: pathlib.Path = None,
     existing files and folder, which fields should be stripped from the model type represented by the path passed in as
     a parameter.
     """
+    # FIXME: set default value of path to Path.cwd()
```
Comment (butler54) [inline trestle/utils/fs.py:158]: I think this is not a fixme. (I'll edit it quickly).

**Review by butler54 — APPROVED** [review]
LGTM

**Pull request merged**